home/farcall.asm

```diff
@@ -2,131 +2,242 @@
 ; Functions that rely on "following" data access it via their return address on
 ; the stack, and update the return address to point past the data.
 
-MACRO update_pushed_af_flags
-; Must have been preceded by 'push af'.
-; Updates the flags on the stack to their current value
-; without clobbering any other registers.
-	push af ; pushes current flags
+_RstFarCall::
+; Call the following dab pointer
+; preserves af, bc, de, hl
+; Partially implemented in the rst handler
+
+; Stack layout
+; +11 return address
+; +10 saved bank
+; +8  return path
+; +6  function pointer
+; +4  saved af
+; +2  saved de
+; +0  saved hl
+.do_farcall
+; same speed and size as add sp, -2 but preserves flags
+	dec sp
+	dec sp
+	push af
+	push de
 	push hl
-	ld hl, sp + 2
-	ld a, [hli] ; reads current flags
-	assert HIGH(wStackBottom) == HIGH(wStackTop)
-	inc l
-	ld [hl], a ; overwrites pushed flags
+
+	ld hl, sp + 10
+	ldh a, [hROMBank]
+	ld [hli], a
+
+; Read return address
+	ld a, [hli]
+	ld h, [hl]
+	ld l, a
+
+; function pointer
+	ld a, [hli]
+	ld e, a
+	ld a, [hli]
+	ld d, a
+
+; bank number
+	ld a, [hli]
+	rst Bankswitch
+
+	push hl
+; write the function pointer
+	ld hl, sp + 9 ; +3 due to the push above and writing d first
+	ld a, d
+	and $7f
+	ld [hld], a
+	ld [hl], e
+
+	xor d
+	pop de
+	jr z, .call
+; for farjp we need to replace the return address
+	ld de, DoNothing
+.call
+; write return address
+	ld hl, sp + 11
+	ld a, e
+	ld [hli], a
+	ld [hl], d
+
+; restore registers and jump to the target function
 	pop hl
+	pop de
 	pop af
-ENDM
-
-RstFarCall::
-; Call the following dba pointer.
-	ldh [hFarCallSavedA], a
-	ld a, h
-	ldh [hFarCallSavedH], a
-	ld a, l
-	ldh [hFarCallSavedL], a
-	pop hl
+	ret
+
+FarPointerCall::
+; Call the dba pointer at hl. Clobbers a and hl.
 	ld a, [hli]
-	ldh [hTempBank], a
+	push af
 	ld a, [hli]
-	bit 7, [hl]
-	jr nz, .farjp
-	inc hl
-	push hl
-	dec hl
-.farjp
 	ld h, [hl]
 	ld l, a
-	res 7, h
-_DoFarCall:
-	ldh a, [hROMBank]
+	pop af
+; fallthrough
+
+FarCall_hl::
+; Call a:hl.
+	dec sp
+	call .do_farcall
+.return
+; restore the rom bank and clean up
 	push af
-	ldh a, [hTempBank]
-_DoFarCall_BankInA:
+	push hl
+	ld hl, sp + 4
+	ld a, [hl]
 	rst Bankswitch
-	call RetrieveAHLAndCallFunction
-_ReturnFarCall:
-	ldh [hFarCallSavedA], a
-	update_pushed_af_flags
+	pop hl
 	pop af
-	rst Bankswitch
-	ldh a, [hFarCallSavedA]
+	inc sp
 	ret
 
-RunFunctionInWRA6::
-; Call the following function in wDecompressScratch's WRAM bank. Clobbers a.
-	ld a, BANK(wDecompressScratch)
-StackCallInWRAMBankA::
-; Call the following function in WRAM bank a. Clobbers a.
-	ldh [hTempBank], a
-	ld a, h
-	ldh [hFarCallSavedH], a
-	ld a, l
-	ldh [hFarCallSavedL], a
+.do_farcall
+	push hl
+	push af
+	ldh a, [hROMBank]
+	ld hl, sp + 6
+	ld [hl], a
+	pop af
+	rst Bankswitch
+; same speed as ret but doesn't leave a stack address in hl
+; less likely to trash the stack if bugs happen
```

Comment:
What's wrong with trashing the stack and crashing if something is this broken?

Reply:
I'd still expect it to crash but the keeping the return address from being trashed would make debugging easier

Reply:
Changed to use call _hl_